### PR TITLE
Adding regularizers

### PR DESCRIPTION
**Pull request by joaogui1**
Right now I added l1 and l2, if there's interest in this PR I will also add the OffDiagonalOrthogonal so we can get the same regularizers as sonnet.

- [ ] OffDiagonalOrthogonal
- [ ] Tests

### CODE REVIEW

File: haiku/_src/regularizers.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    Raises:
+      ValueError: if scale is <0.
+    """
+    if scale < 0:
+        raise ValueError("scale must be a non-negative value")
+    self.scale = scale
+
+
+  def __call__(self, parameters) -> jnp.array:
+    leaves, _ = tree_flatten(parameters)
```
Comment (tomhennigan) [inline haiku/_src/regularizers.py:40]: `jax.tree_leaves` ?

File: haiku/_src/typing.py [hunk 2]
```diff
@@ -36,6 +36,7 @@
 State = Mapping[str, Mapping[str, jnp.ndarray]]
 Padding = Callable[[int], Sequence[int]]
 Paddings = Union[Padding, Sequence[Padding]]
+Regularizer = Callable[[jnp.ndarray], float]
```
Comment (tomhennigan) [inline haiku/_src/typing.py:39]: This takes any tree of ndarrays and returns an ndarray, I think the type should be `Callable[[Any], jnp.ndarray]` or `Callable[[Params], jnp.ndarray]`.

Comment (joaogui1) [reply haiku/_src/typing.py:39]: Yeah, good point!

**Comment by joaogui1** [comment]
I think adding it to JAX would be nice, as it would benefit haiku, flax, trax and all the rest of the JAX ecosystem at once, if @jekbradbury agrees I can write the PR in the JAX repo

**Comment by jekbradbury** [comment]
If we were to add these to JAX, I think we'd rather they be plain functions of pytrees (i.e. not classes, and no `scale` argument). But even then I'm not sure: outside of NumPy/SciPy, where we arguably aren't adding additional API surface for users to learn, we've tried to include only essential abstractions (functions and constructs with implementation subtleties, those which are frequently thought of as black boxes, or those where it's important to match conventions). Even though I agree that regularizers are useful functions, I'm not sure they fall into one of those boxes—WDYT?

**Comment by joaogui1** [comment]
Yeah, JAX is more about functions and I think has no OOP.
I think both regularizations would fit well in `jax.nn` as it seems to have useful functions for neural nets

**Comment by joaogui1** [comment]
Have you guys reached a decision @jekbradbury @tomhennigan ?

**Pull request closed without merging**